crates/rspack_napi/src/threadsafe_function/with_ref.rs

```diff
@@ -0,0 +1,109 @@
+use std::thread::{self, ThreadId};
+
+use derivative::Derivative;
+use napi::bindgen_prelude::{FromNapiValue, ToNapiValue, TypeName, ValidateNapiValue};
+use napi::{bindgen_prelude::JsValuesTupleIntoVec, Env, JsFunction, Ref};
+use napi::{JsUnknown, NapiRaw, NapiValue};
+
+use super::ThreadsafeFunction;
+use crate::errors::NapiResultExt;
+
+/// Thread safe function type that
+/// - implements `ToNapiValue`, and
+/// - can be synchronously called from any thread, including the node thread
+#[derive(Derivative)]
+#[derivative(Debug)]
+#[derivative(Debug(bound = ""))]
+pub struct ThreadSafeFunctionWithRef<T: 'static, R> {
```

Comment:
I would prefer `JsFunctionRef` and `ThreadsafeFunction`. It's not necessary to create a isomorphic struct and changing `JsFilename` to an `enum` seems enough.
You can cut and paste my old implementation to crate `rspack_napi` and share between crates.

Reply:
Sorry I don't quite understand the comments. Could you elaborate more about avoiding isomorphic tsfn and changing `JsFilename` to an `enum`? Thanks.

Meanwhile here's a more detailed explanation on the current design. Hope it clears things up a little bit.
## Why ThreadSafeFunctionWithRef

The ultimate purpose of `ThreadSafeFunctionWithRef` is for `JsFilenameFn` to be both  `Send + Sync` and non-async.
- Field `tsfn: ThreadsafeFunction<T, R>`  is for blocking calls from non-node threads;
- Field `fn_ref: Ref<()>`  is for blocking calls in the node thread.

so I think **both `ThreadsafeFunction` and  `Ref<()>` are needed, unless we make `JsFilenameFn` async**.

`fn_ref` also make it possible to implement `ToNapiValue`, which is required by `#[napi(object)]`.

I don't think we can replace `Ref<()>` with  `JsFunctionRef` because `JsFunctionRef` can't be [safely dropped from non-node threads](https://github.com/branchseer/rspack/blob/bf1ac34f02b2676ac8806fcda449aa1543b45b04/crates/rspack_fs_node/src/node.rs#L25), whereas `Ref<()>` [can](https://docs.rs/napi/latest/src/napi/js_values/value_ref.rs.html#56-61) as it holds zero refcount in `ThreadSafeFunctionWithRef` (we have tsfn to strongly hold the function).
## Why make function type generic in `JsFilename` and `Filename`

We want to avoid creating tsfn when possible, and want the `render` function to be generic on the error type.
- In [`JsCompilation::get_path` and its family](https://github.com/branchseer/rspack/blob/bf1ac34f02b2676ac8806fcda449aa1543b45b04/crates/rspack_binding_values/src/compilation.rs#L360-L405), filename functions are immediately invoked. Creating tsfn is not needed. `JsFilename<JSFunction>` will suffice. We also need `render`'s error type to be `napi::Error` here,  so that these methods can rethrow if the filename function throws.
- As seen in this PR, in many places template-only filenames are passed to `compilation.get_path` or its family. In these cases, the result should always succeed. Generic function type allows us to use `NoFilenameFn` as the function type, whose error type `Infallible`, which allows us to unwrap without handling errors ([`ResultInfallibleExt::always_ok`](https://github.com/branchseer/rspack/blob/bf1ac34f02b2676ac8806fcda449aa1543b45b04/crates/rspack_util/src/infallible.rs#L10)).

Reply:
Here's my point. The `ThreadsafeFunction` and `JsFunction` are two different kinds of stuff. For `ThreadsafeFunction`, it's ok to be called with thread safety and can be dropped and destroyed on the other thread. However, `JsFunction` is not. Mixing these implementations would cause the program hard to debug in the future. I would recommend you to use either one of them, and not try to make an abstraction around these two things.

Reply:
Ok I now understand your concern. I have removed `ThreadsafeFunctionWithRef` and replaced its usages with vanilla `ThreadsafeFunction` in b15d622957a84c0ebf1066b5e58b43f5a44fe11f.

Now the thread-safe version of `Filename` can't be called on the node thread. Fortunately there is no such case according to the tests, and even if there is, tsfn would [detect and report the deadlock](https://github.com/nodejs/node/commit/45676be3dc27683e0a4cd76f981beecf60ed22fb), so that would be easy to debug.